Add basic vignettes (#51)
mrc-ide/hipercow: vignettes_src/troubleshooting.Rmd

r_output(readLines("mycode.R"))
```

And then you either submit a task, **or** a task that you have
previously submitted gets run (which could happen ages after you
submit it if the cluster is busy).

```{r}
hipercow_environment_create(sources = "mycode.R")
id <- task_create_expr(times2(10))
task_wait(id)
task_status(id)
```

The error here has happened before getting to your code - it is
happening when context loads the source files.  The log makes this
a bit clearer:

```{r}
task_log_show(id)
```

## My task got stuck at `submitted`

(Previous users of `didehpc` may recognise this as being stuck at `PENDING`).

This is the most annoying one, and can happen for many reasons. You can see via the [web interface](mrcdata.dide.ic.ac.uk/hpc/) or the Microsoft cluster tools that your task has failed but `hipercow` is reporting it as pending.  This happens when something has failed during the script that runs *before* any `hipercow` code runs on the cluster.

Things that have triggered this situation in the past:

* An error in the Microsoft cluster tools
* A misconfigured node (sometimes they are missing particular software)
* A networking issue
* Gremlins

> * Network path mapping error

Reviewer: * Running out of disk space at some point after submitting the job

Author: added

* Look in the logs to try and find the reason why the failing tasks are doing so.
* Are there variables in the global R environment on your local computer, that your code relies upon, that won't be present on the cluster? Do you have local R packages or sources loaded which you haven't declared when initialising your context?
* Or any system variables, or other dependencies which enable your task to work locally, but won't be set up on a cluster node?
* Are you referring to any files visible to your local machine, but not on the cluster? Are you referring to `C:` for instance?
* (Rarely:) Are you viewing a cached version of any network storage on your local computer, that has not been synced to the real network storage view that the cluster has?
* Check that you have not run out of disk-space. The Q: quota is normally 15Gb.
* If you are running C code, check for other causes of indeterminate failures, such as uninitialised variables, or array out-of-bounds errors. These are unpredictable errors by nature, but surprisingly often you might get away with it on a local computer, while a cluster node behaves differently.

> * If you are running stochastic code, check that you are *really* using the same random number seeds.

Reviewer: ```suggestion
* If you are running stochastic code, check that you are *really* using the same random number seeds. If you are expecting exactly the same result from the cluster job, check that your code produces the same answer if you repeat it locally.
```

* If you find messages about "Error allocating a vector..." or "std::bad_alloc", then try and work out the memory usage of a single task. Perhaps run it locally with task manager (Windows), or `top` (macOS/Linux) running, and watch to see what the memory usage is. If the task is single-core, consider the total memory used if you run 8 or 16 instances on the same cluster machine. If the total memory exceeds the available, then behaviour will be undefined, and some tasks will likely fail.
* In the above example, note that someone else's memory-hungry combination of tasks may affect your small-memory task, if they run on the same node. We don't enforce any memory limits on tasks, which on the whole, is nice and convenient, but it carries the risk that the above can happen.
* Always check you're not running out disk space. The Q: quota is normally 15Gb.
* Find what node your tasks were running on. If you consistently get errors on one node, but not others, then get in touch with Wes, as we do get node failures from time to time, where the fault is not obvious at first.

# My code is slower on the cluster than running locally!

> * This is expected, especially for single-core tasks. Cluster nodes are aiming to provide bandwidth, rather than linear performance, so a single task may run slower on a cluster node than on your own computer. But the cluster node might be able to run 8 or more such tasks at once, without taking any longer, while you continue using your local computer for local things.

Reviewer: ```suggestion
* This is expected, especially for single-core tasks. Cluster nodes are often aiming to provide larger throughput, rather than better linear performance, so a single task may run slower on a cluster node than on your own computer. But the cluster node might be able to run 16 or more such tasks at once, without taking any longer, while you continue using your local computer for local things.
```

* If that is still insufficient, and you still want to compare timings in this way, then check that the cluster is doing *exactly* the same work as your local computer.

## Asking for help

If you need help, you can ask in the "Cluster" teams channel.  This is better than emailing Rich or Wes directly as they may not have time to respond, or may be on leave.

When asking for help it is really important that you make it as easy as possible for us to help you. This is surprisingly hard to do well, and we would ask that you first take a look at these two short articles:

* [How to ask a good question](https://stackoverflow.com/help/how-to-ask)
* [How to create a minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)

Things we will need to know:

* The contents of `hipercow::hipercow_configuration()`
* What you've tried doing
* The values of any errors (not just that they occurred!)
* Logs of the offending task if you have it

> Too often, we will get requests from people that where we have no information about what was run, what packages or versions are being installed, etc. This means your message sits there until we see it, we'll ask for clarification - that message sits there until you see it, you respond with a little more information, and it may be days until we finally discover the root cause of your problem, by which point we're both quite fed up. We will never complain if you provide "too much" information in a good effort to outline where your problem is.

Reviewer: ```suggestion
Too often, we will get requests for help with no information about what was run, what packages or versions are being installed, etc. This means your message sits there until we see it, we'll ask for clarification - that message sits there until you see it, you respond with a little more information, and it may be days until we finally discover the root cause of your problem, by which point we're both quite fed up. We will never complain if you provide "too much" information in a good effort to outline where your problem is.
```
